libs/hwes/event.go

```diff
@@ -2,29 +2,124 @@ package hwes
 
 import (
 	"encoding/json"
+	"fmt"
 	"github.com/EventStore/EventStore-Client-Go/esdb"
 	"github.com/google/uuid"
+	"strings"
 	"time"
 )
 
 type Event struct {
-	ID        uuid.UUID
-	Type      string
-	Data      []byte
-	Timestamp time.Time
+	EventID       uuid.UUID
+	EventType     string
+	AggregateID   uuid.UUID
+	AggregateType AggregateType
+	Data          []byte
+	Timestamp     time.Time
+	Version       uint64
 }
 
-func NewBaseEvent(eventType string) Event {
+func NewEventBase(aggregate Aggregate, eventType string) Event {
 	return Event{
-		ID:        uuid.New(),
-		Type:      eventType,
-		Timestamp: time.Now().UTC(),
+		EventID:       uuid.New(),
+		EventType:     eventType,
+		AggregateID:   aggregate.GetID(),
+		AggregateType: aggregate.GetType(),
+		Timestamp:     time.Now().UTC(),
 	}
 }
 
+func NewEventBaseWithPayload(aggregate Aggregate, eventType string, payload interface{}) (Event, error) {
+	eventBase := NewEventBase(aggregate, eventType)
+	if err := eventBase.SetJsonData(&payload); err != nil {
+		return Event{}, err
+	}
+	return eventBase, nil
+}
+
+// ResolveAggregateIDAndTypeFromStreamID converts a streamID to an aggregateType and aggregateID
+// See aggregate.GetStreamID
+//
+// # Example
+//
+// StreamID:		task-d9027be3-d00f-4eec-b50e-5f489df20433
+// AggregateType: 	task
+// AggregateID: 	d9027be3-d00f-4eec-b50e-5f489df20433
+func ResolveAggregateIDAndTypeFromStreamID(streamID string) (AggregateType, uuid.UUID, error) {
+	streamIDParts := strings.SplitN(streamID, "-", 2)
+
+	var aggregateTypeStr, aggregateIDStr string
+
+	if len(streamIDParts) == 2 {
+		aggregateTypeStr = streamIDParts[0]
+		aggregateIDStr = streamIDParts[1]
+	} else {
+		return "", uuid.UUID{}, fmt.Errorf("cannot resolve aggregateType and aggregateID from streamID '%s'", streamID)
+	}
+
+	aggregateType := AggregateType(aggregateTypeStr)
+	if aggregateType == "" {
+		return "", uuid.UUID{}, fmt.Errorf("resolved empty aggregateType from streamID '%s'", streamID)
+	}
+
+	aggregateID, err := uuid.Parse(aggregateIDStr)
+	if err != nil {
+		return "", uuid.UUID{}, err
+	}
+
+	return aggregateType, aggregateID, err
+}
+
+// NewEventFromRecordedEvent is a helper function for EventStore.
+// This function transforms esdb.RecordedEvent to hwes.Event.
+// We expect that the StreamID of the aggregate is in the format of "[aggregateType]-[aggregateID]".
```

Comment:
is there a reason to let the user of this library know about stream ids at all? can't we just build/parse it when needed?